source/slang/slang-ir-dll-import.cpp

```diff
@@ -0,0 +1,249 @@
+// slang-ir-dll-import.cpp
+#include "slang-ir-dll-import.h"
+
+#include "slang-ir.h"
+#include "slang-ir-insts.h"
+
+namespace Slang
+{
+
+struct DllImportContext
+{
+    IRModule* module;
+    DiagnosticSink* diagnosticSink;
+
+    SharedIRBuilder sharedBuilder;
+
+    IRFunc* loadDllFunc = nullptr;
+    IRFunc* loadFuncPtrFunc = nullptr;
+    IRFunc* stringGetBufferFunc = nullptr;
+
+    IRFunc* createBuiltinIntrinsicFunc(UInt paramCount, IRType** paramTypes, IRType* resultType, UnownedStringSlice targetIntrinsic)
+    {
+        IRBuilder builder(sharedBuilder);
+        builder.setInsertInto(module->getModuleInst());
+        IRFunc* result = builder.createFunc();
+        builder.setInsertInto(result);
+        auto funcType = builder.getFuncType(paramCount, paramTypes, resultType);
+        builder.setDataType(result, funcType);
+        builder.addTargetIntrinsicDecoration(
+            result, CapabilitySet(CapabilityAtom::CPP), targetIntrinsic);
+        return result;
+    }
+
+    IRFunc* getLoadDllFunc()
+    {
+        if (!loadDllFunc)
+        {
+            IRBuilder builder(sharedBuilder);
+            builder.setInsertInto(module->getModuleInst());
+            IRType* stringType = builder.getStringType();
+            loadDllFunc = createBuiltinIntrinsicFunc(
+                1,
+                &stringType,
+                builder.getPtrType(builder.getVoidType()),
+                UnownedStringSlice("_slang_rt_load_dll($0)"));
+        }
+        return loadDllFunc;
+    }
+
+    IRFunc* getLoadFuncPtrFunc()
+    {
+        if (!loadFuncPtrFunc)
+        {
+            IRBuilder builder(sharedBuilder);
+            builder.setInsertInto(module->getModuleInst());
+
+            IRType* stringType = builder.getStringType();
+            IRType* paramTypes[] = {builder.getPtrType(builder.getVoidType()), stringType};
+            loadFuncPtrFunc = createBuiltinIntrinsicFunc(
+                2,
+                paramTypes,
+                builder.getPtrType(builder.getVoidType()),
+                UnownedStringSlice("_slang_rt_load_dll_func($0, $1)"));
```

Comment:
This seems a bit messy, but I agree that the alternative isn't ideal either.

The "correct" way to handle this would, I believe, be to have a specific IR instruction opcode for the `_slang_rt_load_dll_func` operation, and to simply emit an instruction with that opcode directly instead of emitting a call to a generated builtin function. That approach avoids a lot of the boilerplate here, at the cost of needing to have emit logic for uses of that new instruction (which ideally isn't too bad because it can be put in `CLikeSourceEmitter` and shared).

I like the instruction option better than any alternative I can think of to try to generalize the pattern going on here...